src/ol/events.js

```diff
@@ -0,0 +1,347 @@
+  RESIZE: 'resize',
+  TOUCHSTART: 'touchstart',
+  TOUCHMOVE: 'touchmove',
+  TOUCHEND: 'touchend',
+  WHEEL: 'wheel'
+};
+
+
+/**
+ * @enum {number}
+ * @const
+ */
+ol.events.KeyCode = {
+  LEFT: 37,
+  UP: 38,
+  RIGHT: 39,
+  DOWN: 40
+};
+
+
+/**
+ * Property name on an event target for the listener map associated with the
+ * event target.
+ * @const {string}
+ * @private
+ */
+ol.events.LISTENER_MAP_PROP_ = 'olm_' + ((Math.random() * 1e4) | 0);
+
+
+/**
+ * @typedef {EventTarget|ol.events.EventTarget|
+ *     {addEventListener: function(string, Function, boolean=),
+ *     removeEventListener: function(string, Function, boolean=),
+ *     dispatchEvent: function((ol.events.EventType|string))}}
+ */
+ol.events.EventTargetLike;
+
+
+/**
+ * Key to use with {@link ol.Observable#unByKey}.
+ *
+ * @typedef {ol.events.ListenerObjType|Array.<ol.events.ListenerObjType>}
+ * @api
+ */
+ol.events.Key;
+
+
+/**
+ * Listener function. This function is called with an event object as argument.
+ * When the function returns `false`, event propagation will stop.
+ *
+ * @typedef {function(ol.events.Event)|function(ol.events.Event): boolean}
+ * @api
+ */
+ol.events.ListenerFunctionType;
+
+
+/**
+ * @typedef {{bindTo: (Object|undefined),
+ *     boundListener: (ol.events.ListenerFunctionType|undefined),
+ *     callOnce: boolean,
+ *     listener: ol.events.ListenerFunctionType,
+ *     target: (EventTarget|ol.events.EventTarget),
+ *     type: (ol.events.EventType|string),
+ *     useCapture: boolean}}
+ */
+ol.events.ListenerObjType;
+
+
+/**
+ * @param {ol.events.ListenerObjType} listenerObj Listener object.
+ * @return {ol.events.ListenerFunctionType} Bound listener.
+ */
+ol.events.bindListener_ = function(listenerObj) {
+  var boundListener = function(evt) {
+    var rv = listenerObj.listener.call(listenerObj.bindTo, evt);
+    if (listenerObj.callOnce) {
+      ol.events.unlistenByKey(listenerObj);
+    }
+    return rv;
+  }
+  listenerObj.boundListener = boundListener;
+  return boundListener;
+};
+
+
+/**
+ * Finds the matching {@link ol.events.ListenerObjType} in the given listener
+ * array.
+ *
+ * @param {!Array<!ol.events.ListenerObjType>} listenerArray Array of listeners.
+ * @param {!Function} listener The listener function.
+ * @param {boolean} useCapture The capture flag for the listener.
+ * @param {Object=} opt_this The `this` value inside the listener.
+ * @param {boolean=} opt_remove Remove the found listener from the array.
+ * @return {ol.events.ListenerObjType|undefined} The matching listener.
+ * @private
+ */
+ol.events.findListener_ = function(
+    listenerArray, listener, useCapture, opt_this, opt_remove) {
+  var listenerObj;
+  for (var i = 0, ii = listenerArray.length; i < ii; ++i) {
+    listenerObj = listenerArray[i];
+    if (listenerObj.listener === listener &&
+        listenerObj.useCapture == useCapture &&
+        listenerObj.bindTo === opt_this) {
+      if (opt_remove) {
+        listenerArray.splice(i, 1);
```

Comment:
True, but there is also a performance test hinting that splice may be slow. This is not super important though